server.js

```diff
@@ -1,20 +1,119 @@
 const express = require("express");
 const cors = require("cors");
-
+const moment = require("moment"); // require
+moment().format("YYYY-MM-DD");
 const app = express();
 
 app.use(express.json());
 app.use(cors());
 
 //Use this array as your (in-memory) data store.
 const bookings = require("./bookings.json");
+const { body, validationResult } = require("express-validator");
 
 app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
 
 // TODO add your routes and helper functions here
+app.post(
+  "/bookings",
+  [body("email").isEmail().withMessage("Please input a valid email")],
+  function (request, response) {
+    const errors = validationResult(request);
+    if (!errors.isEmpty()) {
+      return response.status(422).json({ errors: errors.array() });
+    }
+    const expectedKeys = [
+      "title",
+      "firstName",
+      "surname",
+      "email",
+      "roomId",
+      "checkInDate",
+      "checkOutDate",
+    ];
+    for (let item in request.body) {
+      if (!request.body[item] || !expectedKeys.includes(item)) {
+        response
+          .status(400)
+          .json("Please ensure your are passing the correct fileds");
+        return;
+      }
+    }
+    if (!moment(request.body.checkOutDate).isAfter(request.body.checkInDate)) {
+      response.status(422).json("Checkout date must be after checkin date");
+      return;
+    }
+    let newBooking = {
+      id: bookings[bookings.length - 1].id + 1,
+      title: request.body.title,
+      firstName: request.body.firstName,
+      surname: request.body.surname,
+      email: request.body.email,
+      roomId: request.body.roomId,
+      checkInDate: request.body.checkInDate,
+      checkOutDate: request.body.checkOutDate,
+    };
+    bookings.push(newBooking);
+    response.json(bookings);
+  }
+);
+
+app.get("/bookings", function (request, response) {
+  response.send(bookings);
+});
+
+app.get("/bookings/search", function (request, response) {
+  let searchTerm = request.query.term;
+  let date = request.query.date;
+  if (moment(date, "YYYY-MM-DD", true).isValid()) {
+    let results = bookings.filter(
+      (booking) =>
+        date === booking.checkInDate ||
+        moment(date).isBetween(booking.checkInDate, booking.checkOutDate)
+    );
+    results.length > 0
+      ? response.json(results)
+      : response.status(404).json("nothing was found on this date");
+  } else if (typeof searchTerm == "string" && searchTerm.trim().length > 0) {
+    const result = bookings.filter(
+      (booking) =>
+        booking.email.toLowerCase().includes(searchTerm.toLowerCase()) ||
+        booking.firstName.toLowerCase().includes(searchTerm.toLowerCase()) ||
+        booking.surname.toLowerCase().includes(searchTerm.toLowerCase())
+    );
+    result.length > 0
+      ? response.json(result)
+      : response.status(404).json("Nothing matched this term");
```

Comment:
People disagree on this, but I think that 404 should be returned when you use a parameter in the path, i.e. bookings/1 and there are no results, because that is the address of a resource which the user expects to exist. However, if you use a query parameter, there isn't necessarily an expectation that the resource exists, so a 200, or 204 is a better response as strictly speaking, nothing has gone wrong. However. that's not a definitive answer, just my opinion. If you like, do some googling, or ask your classmates and come to your own conclusion.

Reply:
This makes sense, I'll use the 204 code for this.